minio/minio
# Heal buckets at node level

## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
Healing buckets at erasure set level used to make lot of calls cross nodes. This change would heal buckets at node level.
It also takes care of removing any dangling buckets as part of this heal.

## Motivation and Context


## How to test this PR?
Follow below steps to verify the haling/removal of dangling buckets

Step-1: Start a 4 node, 1 disk each MinIO setup as below
```
minio server http://127.0.0.1:9000/tmp/bkt-test/1/ http://127.0.0.1:9001/tmp/bkt-test/2/ http://127.0.0.1:9002/tmp/bkt-test/3/ http://127.0.0.1:9003/tmp/bkt-test/4/

minio server http://127.0.0.1:9000/tmp/bkt-test/1/ http://127.0.0.1:9001/tmp/bkt-test/2/ http://127.0.0.1:9002/tmp/bkt-test/3/ http://127.0.0.1:9003/tmp/bkt-test/4/ --address :9001

minio server http://127.0.0.1:9000/tmp/bkt-test/1/ http://127.0.0.1:9001/tmp/bkt-test/2/ http://127.0.0.1:9002/tmp/bkt-test/3/ http://127.0.0.1:9003/tmp/bkt-test/4/ --address :9002

minio server http://127.0.0.1:9000/tmp/bkt-test/1/ http://127.0.0.1:9001/tmp/bkt-test/2/ http://127.0.0.1:9002/tmp/bkt-test/3/ http://127.0.0.1:9003/tmp/bkt-test/4/ --address :9003
```

Step-2: Create two buckets `bkt1` and `bkt2` and load objects to them
```
$ mc alias set myminio http://127.0.0.1:9000 minioadmin minioadmin
$ mc mb myminio/bkt1
$ mc mb myminio/bkt2
$ mc cp /etc/hosts myminio/bkt2
```

Step-3: Now bring down the last MinIO instance and remove first bucket
`mc rb myminio/bkt1 --force`

Step-4: Verify that last disk only contains the stale bucket xl.meta
```
$ ll /tmp/bkt-test/1 
total 4
drwxr-xr-x. 1 shubhendu shubhendu 88 Oct 27 11:23 .minio.sys
drwxr-xr-x. 1 shubhendu shubhendu  8 Oct 27 11:23 ..
drwxr-xr-x. 1 shubhendu shubhendu 10 Oct 27 11:24 bkt2
drwxr-xr-x. 1 shubhendu shubhendu 28 Oct 27 11:25 .

$ ll /tmp/bkt-test/2 
total 4
drwxr-xr-x. 1 shubhendu shubhendu 88 Oct 27 11:23 .minio.sys
drwxr-xr-x. 1 shubhendu shubhendu  8 Oct 27 11:23 ..
drwxr-xr-x. 1 shubhendu shubhendu 10 Oct 27 11:24 bkt2
drwxr-xr-x. 1 shubhendu shubhendu 28 Oct 27 11:25 .

$ ll /tmp/bkt-test/3
total 4
drwxr-xr-x. 1 shubhendu shubhendu 88 Oct 27 11:23 .minio.sys
drwxr-xr-x. 1 shubhendu shubhendu  8 Oct 27 11:23 ..
drwxr-xr-x. 1 shubhendu shubhendu 10 Oct 27 11:24 bkt2
drwxr-xr-x. 1 shubhendu shubhendu 28 Oct 27 11:25 .

$ ll /tmp/bkt-test/4
total 4
drwxr-xr-x. 1 shubhendu shubhendu  8 Oct 27 11:23 ..
drwxr-xr-x. 1 shubhendu shubhendu 36 Oct 27 11:24 .
drwxr-xr-x. 1 shubhendu shubhendu 10 Oct 27 11:24 bkt2
drwxr-xr-x. 1 shubhendu shubhendu 10 Oct 27 11:24 bkt1 <<<<<<<<========== HERE
drwxr-xr-x. 1 shubhendu shubhendu 72 Oct 27 11:24 .minio.sys
```

Step-5: Now start the last MinIO instance back and wait for few secs. Verify that last disk also doesn't contain the removed disk xl.meta anymore
```
$ ll /tmp/bkt-test/4 
total 4
drwxr-xr-x. 1 shubhendu shubhendu  8 Oct 27 11:23 ..
drwxr-xr-x. 1 shubhendu shubhendu 10 Oct 27 11:24 bkt2
drwxr-xr-x. 1 shubhendu shubhendu 72 Oct 27 11:25 .minio.sys
drwxr-xr-x. 1 shubhendu shubhendu 28 Oct 27 11:25 .
```

Without the fix, last disk retains the traces of removed bucket (while 4th MinIO instance was down). This fix as well will remove any dangling bucket which don't contain any data. If bucket contained data, it just logs as error once that dangling bucket cleanup didn't happen as bucket was not empty
```
While deleting dangling Bucket (bkt1), Drive 127.0.0.1:9003:/home/shubhendu/work-forked-repos/minio/bkt-test/4 returned an error (volume is not empty) (*fmt.wrapError
)
```


## Types of changes

## Checklist:
